### PR TITLE
feat: add on investment account provision requested handler

### PR DESCRIPTION
**Pull request by beniutek**
AC-2323 https://getdreams.atlassian.net/browse/AC-2323

add separate method for investment account provisioning
update docs
add tests

### CODE REVIEW

**Review by thelitek — CHANGES_REQUESTED** [review]

**Comment by beniutek** [comment]
I've also regenerated docs, it's all autogenerated based on the comments inside the code.

File: README.md [hunk 1]
```diff
@@ -61,6 +61,47 @@ For more info about building packages refer to [rollup webpage](https://rollupjs
...
+      window.location.href = "http://example.com/some/path"
+    }
+  }
+
+  var sdk = new DreamsWebSDK("https://dreams.api.endpoint");
+  var messageHandler = sdk.setup(callbacks);
+
+  sdk.start(user_jwk_token_value, "en");
+
+  messageHandler.navigateTo('marketplace)
```
Comment (loffek) [inline README.md:90]: Missing the last quote => `'marketplace'`

File: README.md [hunk 2]
```diff
@@ -61,6 +61,47 @@ For more info about building packages refer to [rollup webpage](https://rollupjs
...
+    },
+    onExitRequested: () => {
+      window.location.href = "http://example.com/some/path"
+    }
+  }
+
+  var sdk = new DreamsWebSDK("https://dreams.api.endpoint");
+  var messageHandler = sdk.setup(callbacks);
+
+  sdk.start(user_jwk_token_value, "en");
```
Comment (loffek) [inline README.md:88]: do we need all of this setup to make the example of sending navigateTo? Could the example focus on the different message types instead?

File: README.md [hunk 3]
```diff
@@ -61,6 +61,47 @@ For more info about building packages refer to [rollup webpage](https://rollupjs
...
+```
+
+### message types
+
+currently we can:
+
+- `postUpdateToken` a response to `onIdTokenDidExpire` message
+- `postAccountProvisionInitiated` a response to `onAccountProvisionRequested` message
+- `postInvestmentAccountProvisionInitiated` a response to `onInvestmentAccountProvisionRequested` message
+- `navigateTo` that has no corresponidng dreams side message
```
Comment (loffek) [inline README.md:101]: => corresponding

File: src/events.ts [hunk 4]
```diff
@@ -34,6 +35,10 @@ type AccountProvisionRequestedEvent =  Message & {
   event: 'onAccountProvisionRequested';
 }
 
+type InvestmentAccountProvisionRequestedEvent =  Message & {
+  event: 'onInvestmentAccountProvisionRequested';
```
Comment (loffek) [inline src/events.ts:39]: Don't we need the `externalDreamId` field here? Or what is referred to as `investment account uuid` in the ticket: https://getdreams.atlassian.net/browse/AC-2323

**Review by loffek — CHANGES_REQUESTED** [review]
Good stuff! Some 💅  and a question on the event payload though:

File: src/events.ts (hunk 4 above)
Comment (beniutek) [reply src/events.ts:39]: good point!

File: src/events.ts [hunk 5]
```diff
@@ -53,7 +62,44 @@ type DreamsEvent = IdTokenDidExpireEvent |
                    ExitRequestedEvent |
                    ShareEvent;
 
+type AccountProvisionInitiatedEvent = {
+  event: 'accountProvisioned';
```
Comment (thelitek) [inline src/events.ts:66]: Shouldn't it be `messages.accountProvisionInitiated`? If so shouldn't `messages` be renamed to `event`?

File: src/events.ts (hunk 4 above)
Comment (beniutek) [reply src/events.ts:41]: it's fixed now, let me know if it makes sense

File: src/events.ts (hunk 5 above)
Comment (beniutek) [reply src/events.ts]: good point, renamed it :)

File: src/events.ts [hunk 6]
```diff
@@ -1,41 +1,50 @@
...
 }
 
-type MessageEvent = Message & {
-  event: messages;
+type TokenDidExpireMessage = Message & {
+  idToken: string;
 }
 
-type IdTokenDidExpireEvent =  Message & {
+type InvestmentAccountProvisionMessage = Message & {
```
Comment (loffek) [inline src/events.ts:25]: To be consistent with other naming: => `InvestmentAccountProvisionRequestedMessage` ?

File: src/events.ts [hunk 7]
```diff
@@ -1,41 +1,50 @@
...
+  accountId: string;
+  dreamId: string;
```
Comment (loffek) [inline src/events.ts:27]: Please remind me, is `dreamId` necessary to expose?

File: src/messageHandler.ts [hunk 8]
```diff
@@ -48,36 +66,61 @@ export class MessageHandler {
...
+      name: partnerEvents.accountProvisionInitiated,
+      message
+    }
+
+    this.postMessage(event);
+  }
+
+  /**
+  * You can use this method if you need to manually inform the dreams app that investment account provision has been initiated.
+  * DreamId should match the one received initially with the onInvestmentAccountProvisionRequested event.
```
Comment (loffek) [inline src/messageHandler.ts:92]: needed?

**Review by loffek — CHANGES_REQUESTED** [review]
Only small stuff. Except perhaps the `dreamId` which would affect versioning if we change it later.

File: src/events.ts (hunk 7 above)
Comment (beniutek) [reply src/events.ts:27]: well, how we will connect account with the dream once the provisioning is done? I think we have to expose it

File: src/messageHandler.ts (hunk 8 above)
Comment (beniutek) [reply src/messageHandler.ts:92]: check my comment above -> IMO we need to expose it. Happy to have a chat about it though

File: README.md [hunk 9]
```diff
@@ -61,6 +61,49 @@ For more info about building packages refer to [rollup webpage](https://rollupjs
...
+
+      return data.token;
+    },
+    onAccountProvisionRequested: () => {
+      await fetch("/provision-account-endpoint")
+    },
+    onExitRequested: () => {
+      window.location.href = "http://example.com/some/path"
+    }
+  }
```
[the remark this answers is not in the document]
Comment (beniutek) [reply README.md:83]: Right, will remove that

File: src/events.ts (hunk 7 above)
Comment (loffek) [reply src/events.ts:27]: Ah. The partner needs to call `POST /customers/<customer_id>/dreams/<dreamId>/account-provisioned` ?
But in that case, why are we sending the `accountId`?

I'm thinking we could store a tentative `accountId` on the dream, and then have an endpoint like:

```
POST /customers/<customer_id>/investment-accounts/accountId/provisioned
```

And then we dig up the dream with the corresponding `accountId`.
Or add an optional parameter to the old endpoint:

```
POST /customers/<customer_id>/account-provisioned
{
  accountId: string // defaults to the user savings account.
}
```

I think storing the accountId on the dream can have other benefits as well. E.g. we can always send the same id to the partner to make account provisioning idempotent.
WDYT?

Comment (beniutek) [reply src/events.ts:27]: That does sound sound. I'll check in DES how the endpoint works and make changes

Comment (beniutek) [reply src/events.ts]: Removed `dreamId` let me know if that looks okay now

File: src/events.ts [hunk 10]
```diff
@@ -44,15 +56,64 @@ type ExitRequestedEvent = {
...
+  message: {
+    requestId: string;
+  }
+}
+
+type InvestmentAccountProvisionInitiatedEvent = {
+  name: partnerEvents.investmentAccountProvisionInitiated;
+  message: {
+    requestId: string;
+    dreamId?: string;
```
Comment (loffek) [inline src/events.ts:74]: :trollface:

File: src/events.ts (hunk 7 above)
Comment (loffek) [reply src/events.ts]: Perfect!

File: src/events.ts (hunk 10 above)
Comment (beniutek) [reply src/events.ts]: duh! I'm going to move from js tests to ts tests (but as a separate effort - this ticket is a bit out of hand already)

**Review by loffek — DISMISSED** [review]
🚢

**Review by loffek — APPROVED** [review]

**Pull request merged**